plugins/out_azure_kusto/azure_kusto.c

```diff
@@ -0,0 +1,509 @@
+#include "azure_kusto_conf.h"
+#include "azure_kusto_ingest.h"
+
+/* Create a new oauth2 context and get a oauth2 token */
+static int azure_kusto_get_oauth2_token(struct flb_azure_kusto *ctx)
+{
+    int ret;
+    char *token;
+
+    /* Clear any previous oauth2 payload content */
+    flb_oauth2_payload_clear(ctx->o);
+
+    ret = flb_oauth2_payload_append(ctx->o, "grant_type", 10, "client_credentials", 18);
+    if (ret == -1) {
+        flb_plg_error(ctx->ins, "error appending oauth2 params");
+        return -1;
+    }
+
+    ret = flb_oauth2_payload_append(ctx->o, "scope", 5, FLB_AZURE_KUSTO_SCOPE, -1);
+    if (ret == -1) {
+        flb_plg_error(ctx->ins, "error appending oauth2 params");
+        return -1;
+    }
+
+    ret = flb_oauth2_payload_append(ctx->o, "client_id", -1, ctx->client_id, -1);
+    if (ret == -1) {
+        flb_plg_error(ctx->ins, "error appending oauth2 params");
+        return -1;
+    }
+
+    ret = flb_oauth2_payload_append(ctx->o, "client_secret", -1, ctx->client_secret, -1);
+    if (ret == -1) {
+        flb_plg_error(ctx->ins, "error appending oauth2 params");
+        return -1;
+    }
+
+    /* Retrieve access token */
+    token = flb_oauth2_token_get(ctx->o);
+    if (!token) {
+        flb_plg_error(ctx->ins, "error retrieving oauth2 access token");
+        return -1;
+    }
+
+    return 0;
+}
+
+flb_sds_t get_azure_kusto_token(struct flb_azure_kusto *ctx)
+{
+    int ret = 0;
+    flb_sds_t output = NULL;
+
+    if (pthread_mutex_lock(&ctx->token_mutex)) {
+        flb_plg_error(ctx->ins, "error locking mutex");
+        return NULL;
+    }
+
+    if (flb_oauth2_token_expired(ctx->o) == FLB_TRUE) {
+        ret = azure_kusto_get_oauth2_token(ctx);
+    }
+
+    /* Copy string to prevent race conditions (get_oauth2 can free the string) */
+    if (ret == 0) {
+        output = flb_sds_create(ctx->o->token_type);
+        flb_sds_printf(&output, " %s", ctx->o->access_token);
+    }
+
+    if (pthread_mutex_unlock(&ctx->token_mutex)) {
+        flb_plg_error(ctx->ins, "error unlocking mutex");
+        if (output) {
+            flb_sds_destroy(output);
+        }
+        return NULL;
+    }
+
+    return output;
+}
+
+/**
+ * Executes a control command against kusto's endpoint
+ *
+ * @param ctx       Plugin's context
+ * @param csl       Kusto's control command
+ * @return flb_sds_t      Returns the response or NULL on error.
+ */
+flb_sds_t execute_ingest_csl_command(struct flb_azure_kusto *ctx, const char *csl)
+{
+    flb_sds_t token;
+    flb_sds_t body;
+    size_t b_sent;
+    int ret;
+    struct flb_upstream_conn *u_conn;
+    struct flb_http_client *c;
+    flb_sds_t resp;
+
+    /* Get upstream connection */
+    u_conn = flb_upstream_conn_get(ctx->u);
+    if (!u_conn) {
+        return NULL;
+    }
+
+    /* Get or renew Token */
+    token = get_azure_kusto_token(ctx);
+    if (!token) {
+        flb_plg_error(ctx->ins, "cannot retrieve oauth2 token");
+        goto execute_error;
```

Comment:
At this point neither `c` nor `body` have been initialized so this jump causes a check that could return a false positive and call the destructors for the http client or sds depending on the contents of the stack before this function is called.

In general I think if not strictly necessary `goto` should not be used but if that's not an option then please initialize those variables to NULL at the beginning of the function.

Reply:
Done, removed all gotos except those in azure_kusto_load_ingestion_resources, I think goto is helpful with handling mutex unlock on errors

Reply:
Done, removed all gotos except those in azure_kusto_load_ingestion_resources, I think goto is helpful with handling mutex unlock on errors